### PR TITLE
Fixes #26727 - Fix issue with primary interface on bond

### PR DESCRIPTION
**Pull request by ofalk**
Prerequisites for testing:
* Host with bond interface as primary interface

Tests needed (anticipated impacts):
* Edit host with bond

Steps to Reproduce:
1. create a host with a bond0 interface mode=0 with the ip address and add one slave (enough to test) 
2. delete the host in foreman 
3. run puppet to upload the facts, this should create through the puppet ENC the host in foreman
4. check the interfaces of the host, the bond0 is missing

We have this issue open with several customers since quite a while; Therefore I'm now proposing this directly here. Please let me know if something needs to be adapted.

### CODE REVIEW

**Comment by ofalk** [comment]
OK from my point of view now. Please have a look!

**Comment by tbrisker** [comment]
ok to test

**Comment by mmoll** [comment]
@ofalk Please rebase and have a look at the test failures.

**Comment by ofalk** [comment]
Merged in latest develop branch, but need to look at the tests. Will get back to you as soon as possible!

**Comment by lzap** [comment]
Hello @ofalk! Can you please upload `facter --json` output of the host you are testing with? I would use that for my testing.

**Comment by lzap** [comment]
What you are doing is a big change in behavior and also suggested interface method which is used by plugins. I do not understand the misbehavior, can you first specify how Foreman behaves today and what is the actual bug, then how you want to fix it?

**Comment by ofalk** [comment]
Hey @lzap , what this should fix is the following: https://bugzilla.redhat.com/1503059
This RHBZ is around since quite a while already, but nobody took care, so I moved the issue upstream now. I'm not that good in Ruby, so fixing it could be pretty hard for me. :-(

**Comment by ofalk** [comment]
@lzap in RHBZ#1503059, comment 12, you'll also find some facter output from a test box...

**Comment by timogoebel** [comment]
> @lzap in RHBZ#1503059, comment 12, you'll also find some facter output from a test box...

I just wanted to take a look, but this seems to be a private comment.
We definitely need a proper test for this to make sure we don't break it again in the future. The more real the fact data is, the better. Ideally, we'll add the json (in an anonymizes form) as a static fixture verify the fact importer creates a host with the proper interfaces.
We should also make sure this is correctly processed for a host that already exists. If a user manually configures a bond on a host, foreman should correctly display the interfaces after a fact upload.

File: app/services/fact_parser.rb [hunk 1]
```diff
@@ -79,10 +79,10 @@ def interfaces
   end
 
   # tries to detect primary interface among interfaces using host name
-  def suggested_primary_interface(host)
+  def suggested_primary_interface(hname)
```
Comment (timogoebel) [inline app/services/fact_parser.rb:82]: Nit: please use `hostname` here, nobody will know what `hname` means.

In addition, this breaks the API which a lot of plugins use. :-(

**Comment by ofalk** [comment]
> > @lzap in RHBZ#1503059, comment 12, you'll also find some facter output from a test box...
> 
> I just wanted to take a look, but this seems to be a private comment.
> We definitely need a proper test for this to make sure we don't break it again in the future. The more real the fact data is, the better. Ideally, we'll add the json (in an anonymizes form) as a static fixture verify the fact importer creates a host with the proper interfaces.
> We should also make sure this is correctly processed for a host that already exists. If a user manually configures a bond on a host, foreman should correctly display the interfaces after a fact upload.

From the comment 12:

```
# facter 
interfaces => bond0,eth0,eth1,lo
ipaddress => 172.29.0.95
ipaddress_bond0 => 172.29.0.95
ipaddress_lo => 127.0.0.1
macaddress => 52:54:00:c4:3b:20
macaddress_bond0 => 52:54:00:c4:3b:20
macaddress_eth0 => 52:54:00:C4:3B:20
macaddress_eth1 => 52:54:00:25:12:CF
mtu_bond0 => 1500
mtu_eth0 => 1500
mtu_eth1 => 1500
mtu_lo => 65536
netmask => 255.255.255.0
netmask_bond0 => 255.255.255.0
netmask_lo => 255.0.0.0
network_bond0 => 172.29.0.0
network_lo => 127.0.0.0
```

But I'm organizing the output of the real machine as well!

**Comment by ofalk** [comment]
Hi!

Here is a real life example:
```
$ sudo facter --json | jq .networking
{
  "domain": „xx.xxxx.tld“,
  "fqdn": „xx-xx-9999.xx.xxxx.tld“,
  "hostname": „xx-xx-9999“,
  "interfaces": {
    "bond0": {
      "bindings": [
        {
          "address": "10.92.14.5",
          "netmask": "255.255.255.0",
          "network": "10.92.14.0"
        }
      ],
      "ip": "10.92.14.5",
      "mac": "00:e0:ed:73:54:af",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "10.92.14.0"
    },
    "bond1": {
      "bindings": [
        {
          "address": "10.92.65.5",
          "netmask": "255.255.255.0",
          "network": "10.92.65.0"
        }
      ],
      "ip": "10.92.65.5",
      "mac": "00:e0:ed:72:18:a5",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "10.92.65.0"
    },
    "bond2": {
      "bindings": [
        {
          "address": "192.168.0.2",
          "netmask": "255.255.255.0",
          "network": "192.168.0.0"
        }
      ],
      "ip": "192.168.0.2",
      "mac": "00:e0:ed:74:01:60",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "192.168.0.0"
    },
    "bond3": {
      "bindings": [
        {
          "address": "10.92.129.28",
          "netmask": "255.255.255.0",
          "network": "10.92.129.0"
        }
      ],
      "ip": "10.92.129.28",
      "mac": "00:e0:ed:72:18:a4",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "10.92.129.0"
    },
    "enP3p1s0f0": {
      "mac": "00:e0:ed:72:18:a4",
      "mtu": 9000
    },
    "enP3p1s0f1": {
      "mac": "00:e0:ed:72:18:a5",
      "mtu": 9000
    },
    "enP3p3s0f0": {
      "mac": "00:e0:ed:73:54:af",
      "mtu": 9000
    },
    "enP3p3s0f1": {
      "bindings": [
        {
          "address": "192.168.2.2",
          "netmask": "255.255.255.0",
          "network": "192.168.2.0"
        }
      ],
      "ip": "192.168.2.2",
      "mac": "00:e0:ed:73:54:b0",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "192.168.2.0"
    },
    "enP3p4s0f0": {
      "mac": "00:e0:ed:74:01:60",
      "mtu": 9000
    },
    "enP3p4s0f1": {
      "mac": "00:e0:ed:74:01:61",
      "mtu": 9000
    },
    "enp7s0": {
      "mac": "08:00:69:17:b3:0e",
      "mtu": 1500
    },
    "ens1f0": {
      "mac": "00:e0:ed:7d:68:32",
      "mtu": 9000
    },
    "ens1f1": {
      "bindings": [
        {
          "address": "192.168.1.2",
          "netmask": "255.255.255.0",
          "network": "192.168.1.0"
        }
      ],
      "ip": "192.168.1.2",
      "mac": "00:e0:ed:7d:68:33",
      "mtu": 9000,
      "netmask": "255.255.255.0",
      "network": "192.168.1.0"
    },
    "ens2f0": {
      "mac": "00:e0:ed:74:01:64",
      "mtu": 9000
    },
    "ens2f1": {
      "mac": "00:e0:ed:74:01:65",
      "mtu": 9000
    },
    "lo": {
      "bindings": [
        {
          "address": "127.0.0.1",
          "netmask": "255.0.0.0",
          "network": "127.0.0.0"
        }
      ],
      "ip": "127.0.0.1",
      "mtu": 65536,
      "netmask": "255.0.0.0",
      "network": "127.0.0.0"
    }
  },
  "ip": "10.92.14.5",
  "mac": "00:e0:ed:73:54:af",
  "mtu": 9000,
  "netmask": "255.255.255.0",
  "network": "10.92.14.0",
  "primary": "bond0"
}
```

**Comment by ofalk** [comment]
Note, we have also applied the patch from here:
    https://bugzilla.redhat.com/show_bug.cgi?id=1528193

**Comment by ofalk** [comment]
Hey. Didn't hear anything back from you guys. Is someone able to work on this?
Thanks,
 Oliver

**Comment by tbrisker** [comment]
Hi @ofalk, please take a look at the review from @timogoebel - https://github.com/theforeman/foreman/pull/6731#pullrequestreview-245527867. 
IIUC, looks like this is missing some tests as well as breaks the API of a method used by plugins. Also, looks like tests are failing for this PR - you can click on the "details" link next to the check to see more information. Please address these issue and rebase on the latest develop branch.

**Comment by timogoebel** [comment]
@ofalk: I've filed a PR with a slightly modified fix + test coverage at #6933. Please let me know if this works for you.

**Comment by ofalk** [comment]
> @ofalk: I've filed a PR with a slightly modified fix + test coverage at #6933. Please let me know if this works for you.

Thanks Timo - I'll check it and let you know as soon as possible!

**Comment by lzap** [comment]
And I merged the PR, thanks!

**Pull request closed without merging**